polkadot/node/subsystem-util/src/lib.rs

```diff
@@ -440,3 +442,210 @@ impl Validator {
+
+			let chunk_indices = shuffle
+				.clone()
+				.into_iter()
+				.cycle()
+				.skip(core_start_index)
+				.take(shuffle.len())
+				.collect();
+
+			Some(chunk_indices)
+		} else {
+			None
+		}
+	}
+
+	/// Return and populate the cache with the per-validator chunk index.
+	/// Should only be called if `query_cache_for_validator` returns `None`.
+	pub fn populate_for_validator(
+		&mut self,
+		maybe_client_features: Option<ClientFeatures>,
+		babe_randomness: BabeRandomness,
+		n_validators: usize,
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+		validator_index: ValidatorIndex,
+	) -> ChunkIndex {
+		let shuffle = Self::get_shuffle(
+			maybe_client_features.as_ref(),
+			block_number,
+			babe_randomness,
+			n_validators,
+		);
+		self.0.insert((block_number, session_index), (shuffle, maybe_client_features));
+
+		self.query_cache_for_validator(block_number, session_index, para_id, validator_index)
+			.expect("We just inserted the entry.")
+	}
+
+	/// Return and populate the cache with the per-para chunk index vector.
+	/// Should only be called if `query_cache_for_para` returns `None`.
+	pub fn populate_for_para(
+		&mut self,
+		maybe_client_features: Option<ClientFeatures>,
+		babe_randomness: BabeRandomness,
+		n_validators: usize,
+		block_number: BlockNumber,
+		session_index: SessionIndex,
+		para_id: ParaId,
+	) -> Vec<ChunkIndex> {
+		let shuffle = Self::get_shuffle(
+			maybe_client_features.as_ref(),
+			block_number,
+			babe_randomness,
+			n_validators,
+		);
+		self.0.insert((block_number, session_index), (shuffle, maybe_client_features));
+
+		self.query_cache_for_para(block_number, session_index, para_id)
+			.expect("We just inserted the entry.")
+	}
+
+	fn get_shuffle(
+		maybe_client_features: Option<&ClientFeatures>,
+		block_number: BlockNumber,
+		mut babe_randomness: BabeRandomness,
+		n_validators: usize,
+	) -> Vec<ChunkIndex> {
+		let mut indices: Vec<_> = (0..n_validators)
+			.map(|i| ChunkIndex(u32::try_from(i).expect("validator count should not exceed u32")))
+			.collect();
+
+		if let Some(features) = maybe_client_features {
+			if features.contains(ClientFeatures::AVAILABILITY_CHUNK_SHUFFLING) {
+				let block_number_bytes = block_number.to_be_bytes();
+				for i in 0..32 {
+					babe_randomness[i] ^= block_number_bytes[i % block_number_bytes.len()];
+				}
+
+				let mut rng: ChaCha8Rng = SeedableRng::from_seed(babe_randomness);
+
+				indices.shuffle(&mut rng);
+			}
+		}
+
+		indices
+	}
+
+	/// Return the availability chunk start index for this para.
+	fn para_start_index(
+		maybe_client_features: Option<&ClientFeatures>,
+		n_validators: usize,
+		para_id: ParaId,
+	) -> usize {
+		if let Some(features) = maybe_client_features {
+			if features.contains(ClientFeatures::AVAILABILITY_CHUNK_SHUFFLING) {
+				let mut rng: ChaCha8Rng =
+					SeedableRng::from_seed(
+						u32::from(para_id).to_be_bytes().repeat(8).try_into().expect(
```

Comment:
Nit: What's  the point of doing this if the seed is known then this is not actually random, feels like you want to do some hashing here, don't you ?

Reply:
we need the randomly generated number to be known for a given ParaId (so that all validators have the same numbers), but random between different ParaIds (so that the validators holding the systematic chunks have as little overlap as possible between different paras)